bin/npm.ps1

```diff
@@ -22,11 +22,14 @@ if (Test-Path $NPM_PREFIX_NPM_CLI_JS) {
   $NPM_CLI_JS=$NPM_PREFIX_NPM_CLI_JS
 }
 
+$NPM_ARGS = $MyInvocation.Line.Substring($MyInvocation.InvocationName.Length).Trim()
+$INVOKE_NPM = "$($NODE_EXE -Replace ' ', '` ') $($NPM_CLI_JS -Replace ' ', '` ') $NPM_ARGS".Trim()
```

Comment:
We should be absolutely sure this won't overlap w/ npm's escaping when it spawns scripts.

Reply:
Converted to draft as I’m still experimenting with this.

Reply:
@lukekarrys also please don't use `MyInvocation.Line`, use `MyInvocation.Statement` instead. Here's why: 

If `npm.ps1` is invoked as part of a multi-statement one-liner, like `echo "Hello" && npm start -- commmand --arg=value`, then `MyInvocation` props would look like this:

```text
  "Line": "echo \"Hello\" && npm start -- commmand --arg=value",
  "Statement": "npm start -- commmand --arg=value",
  "InvocationName": "npm",
```

PS Thanks for a shoutout! 🙂

PS C:\temp\cli-test> echo "Hello" && ./ps-test start -- commmand --arg=value
Hello
First Parameter: start
Remaining Arguments: commmand --arg=value
PSCommandPath: C:\temp\cli-test\ps-test.ps1
WARNING: Resulting JSON is truncated as serialization has exceeded the set depth of 2.
MyInvocation: {
  "MyCommand": {
    "Path": "C:\\temp\\cli-test\\ps-test.ps1",
    "Definition": "C:\\temp\\cli-test\\ps-test.ps1",
    "Source": "C:\\temp\\cli-test\\ps-test.ps1",
    "Visibility": 0,
    "ScriptBlock": {
      "Attributes": "",
      "File": "C:\\temp\\cli-test\\ps-test.ps1",
      "IsFilter": false,
      "IsConfiguration": false,
      "Module": null,
      "StartPosition": "System.Management.Automation.PSToken",
      "DebuggerHidden": false,
      "Id": "26129ef5-52a7-4508-903c-65f34757ceae",
      "Ast": "param (\r\n    [string]$firstArg\r\n)\r\n\r\nWrite-Host \"First Parameter: $firstArg\"\r\nWrite-Host \"Remaining Arguments: $args\"\r\nWrite-Host \"PSCommandPath: $PSCommandPath\"\r\nWrite-Host \"MyInvocation: $($MyInvocation | ConvertTo-Json)\"\r\nWrite-Host \"Params:$($MyInvocation.Statement.Substring($MyInvocation.InvocationName.Length))\"\r\n"
    },
    "OutputType": [],
    "ScriptContents": "param (\r\n    [string]$firstArg\r\n)\r\n\r\nWrite-Host \"First Parameter: $firstArg\"\r\nWrite-Host \"Remaining Arguments: $args\"\r\nWrite-Host \"PSCommandPath: $PSCommandPath\"\r\nWrite-Host \"MyInvocation: $($MyInvocation | ConvertTo-Json)\"\r\nWrite-Host \"Params:$($MyInvocation.Statement.Substring($MyInvocation.InvocationName.Length))\"\r\n",
    "OriginalEncoding": {
      "Preamble": null,
      "BodyName": "utf-8",
      "EncodingName": "Unicode (UTF-8)",
      "HeaderName": "utf-8",
      "WebName": "utf-8",
      "WindowsCodePage": 1200,
      "IsBrowserDisplay": true,
      "IsBrowserSave": true,
      "IsMailNewsDisplay": true,
      "IsMailNewsSave": true,
      "IsSingleByte": false,
      "EncoderFallback": "System.Text.EncoderReplacementFallback",
      "DecoderFallback": "System.Text.DecoderReplacementFallback",
      "IsReadOnly": true,
      "CodePage": 65001
    },
    "Name": "ps-test.ps1",
    "CommandType": 16,
    "Version": null,
    "ModuleName": "",
    "Module": null,
    "RemotingCapability": 1,
    "Parameters": {
      "firstArg": "System.Management.Automation.ParameterMetadata"
    },
    "ParameterSets": [
      "[[-firstArg] <string>]"
    ]
  },
  "BoundParameters": {
    "firstArg": "start"
  },
  "UnboundArguments": [
    "commmand",
    "--arg=value"
  ],
  "ScriptLineNumber": 1,
  "OffsetInLine": 17,
  "HistoryId": 1,
  "ScriptName": "",
  "Line": "echo \"Hello\" && ./ps-test start -- commmand --arg=value",
  "Statement": "./ps-test start -- commmand --arg=value",
  "PositionMessage": "At line:1 char:17\r\n+ echo \"Hello\" && ./ps-test start -- commmand --arg=value\r\n+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~",
  "PSScriptRoot": "",
  "PSCommandPath": null,
  "InvocationName": "./ps-test",
  "PipelineLength": 1,
  "PipelinePosition": 1,
  "ExpectingInput": false,
  "CommandOrigin": 0,
  "DisplayScriptPosition": null
}
Params: start -- commmand --arg=value

Reply:
Thanks for the tip @noseratio. I used `MyInvocation.Line` in a force-push after I couldn't get `MyInvocation.Statement` working. In the tests these files are called directly via `child_process.spawn` so I'm not sure if that is affecting them. I will need to do some more research to see why both `Line` and `Statement` are not available on `MyInvocation` in tests.

Reply:
But `MyInvocation.Statement` did work for me locally.

Reply:
I believe, in Windows `npm.ps1` only gets invoked as `npm` if we are already in a PowerShell prompt, or inside another PowerShell script.  Otherwise, it's always `npm.cmd`. So maybe tests should be done via `pwsh.exe`, rather then via `child_process.spawn` directly.

<img width="778" alt="image" src="https://github.com/npm/cli/assets/4774875/e6fa203d-c50b-4c6f-a34b-5efdc80f63a8">

Reply:
Also interestingly, if `npm` is typed in a Git-Bash prompt on Windows, the Bash version - `"C:\Program Files\nodejs\npm"` - will get invoked.

Reply:
Last but not least, `npm.cmd` remains being super important, as it will be invoked for nested npm scripts (like `startapp` below), regardless of the outer shell:

`package.json`:
```json
{
  "name": "cli-test",
  "version": "1.0.0",
  "main": "index.js",
  "scripts": {
    "start": "node index.js",
    "startapp": "npm start -- ",
    "showcli": "echo"
  }
}
```


```text
C:\temp\cli-test>pwsh –noprofile                                                                                                      PowerShell 7.4.2
PS C:\temp\cli-test> npm run startapp -- command --arg=value
Hello from npm.ps1

> cli-test@1.0.0 startapp
> npm start -- command --arg=value
Hello from npm.bat

> cli-test@1.0.0 start
> node index.js command --arg=value

[
  'C:\\Program Files\\nodejs\\node.exe',
  'C:\\temp\\cli-test\\index.js',
  'command',
  '--arg=value'
]
PS C:\temp\cli-test> & "C:\Program Files\Git\bin\bash.exe"

Master@i3msi MINGW64 /c/temp/cli-test
$ npm run startapp -- command --arg=value
Hello from npm[.sh]

> cli-test@1.0.0 startapp
> npm start -- command --arg=value

Hello from npm.bat

> cli-test@1.0.0 start
> node index.js command --arg=value

[
  'C:\\Program Files\\nodejs\\node.exe',
  'C:\\temp\\cli-test\\index.js',
  'command',
  '--arg=value'
]
```

That can probably only be changed by overriding the shell in `.npmrc`:
```ini
shell = "pwsh"
```